megatron/utils.py

```diff
@@ -223,3 +223,29 @@ def print_rank_last(message):
             print(message, flush=True)
     else:
         print(message, flush=True)
+
+
+def get_tflops(batch_size, elapsed_time_per_iteration):
+    """Get tflop/s/GPU from global-batch-size and elapsed-time"""
+    args = get_args()
+    seq_len = args.seq_length
+    hidden_size = args.hidden_size
+    num_layers = args.num_layers
+    vocab_size = args.padded_vocab_size
+
+    # Compute throughput.
+    samples_per_sec = batch_size / elapsed_time_per_iteration
+    tokens_per_sec = samples_per_sec * seq_len
+
+    # General TFLOPs formula (borrowed from Equation 3 in Section 5.1 of
+    # https://arxiv.org/pdf/2104.04473.pdf).
+    # The factor of 4 is when used with activation check-pointing,
+    # otherwise it will be 3, but for 200B model, activation check-pointing will always be on.
+    checkpoint_activations_factor = 4 if args.recompute_granularity == 'full' else 3
+    # GLU activations double the hidden states in the upscaling feed-forward in each transformer layer
+    # This leads to 16bsh^2 instead of 8bsh^2 per first feed-forward layer in MLP, thus we increase the coefficient by 8.
+    # Refer to https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/283#issue-1260805063 for more details.
+    coefficient = 32 if args.glu_activation else 24
+    flops_per_iteration = (coefficient * checkpoint_activations_factor * batch_size * seq_len * num_layers * (hidden_size**2)) * (1. + (seq_len / (6. * hidden_size)) + (vocab_size / (16. * num_layers * hidden_size)))
+    tflops = flops_per_iteration / (elapsed_time_per_iteration * args.world_size * (10**12))
+    return tflops
```

Comment:
Fix formula to take into account MQA
see https://github.com/bigcode-project/Megatron-LM/blob/9ce961178d4dbf920ac9ed0d5443e9ebff0bd6da/megatron/model/transformer.py#L496-L524 for ref :)
```suggestion
    flops_per_iteration = (coefficient * checkpoint_activations_factor * batch_size * seq_len * num_layers * (hidden_size**2)) * (1. + (seq_len / (6. * hidden_size)) + (vocab_size / (16. * num_layers * hidden_size)))
    if args.attention_head_type == 'multiquery':
        # In MultiQuery attention, keys and values are shared across heads
        # qkv projection: 8Bsh^2 -> 4Bsh^2 + 4Bsd_kv
        # The formula in https://arxiv.org/pdf/2104.04473.pdf becomes:
        # 4 * (20 Bsh^2 + 4Bsd_kv + 4Bs^2h) = 96Bsh^2 (1 + (d_kv+s)/6h)
        coefficient = 28 if args.glu_activation else 20 # We substract 4 because of shared kv projection
        d_kv = args.kv_channels
        flops_per_iteration = (coefficient * checkpoint_activations_factor * batch_size * seq_len * num_layers * (hidden_size**2)) * (1. + ((seq_len + d_kv) / (6. * hidden_size)) + (vocab_size / (16. * num_layers * hidden_size)))
    tflops = flops_per_iteration / (elapsed_time_per_iteration * args.world_size * (10**12))
    return tflops
```

Reply:
Good catch, thanks!
Some points seem unclear to me. I assume that in your notation, `d_kv` would be the head-dimension?
I think QKV projection should rather be:
```
# qkv projection: 6Bsh^2 -> 2Bsh^2 + 4Bshd_kv
```
(8 -> 6 and note the additional `h` in the second term)

And the complete formula:
```
# 4 * (20 Bsh^2 + 4Bshd_kv + 4Bs^2h) = 4*20*Bsh^2 (1 + d_kv/5h + s/5h)
```

Reply:
> I assume that in your notation, d_kv would be the head-dimension?

d_kv is kv_channels yes!

> I think QKV projection should rather be:

Yes, good catch! I used `8Bsh^2` because I counted all projections in self attention, but if we talk only about qkv projection it's just `6Bsh^2`

> # 4 * (20 Bsh^2 + 4Bshd_kv + 4Bs^2h) = 4*20*Bsh^2 (1 + d_kv/5h + s/5h)

Good catch! I forgot to update the denominators, and I'm afraid the person who added the case for `args.glu_activation` before me did the same mistake

Reply:
Check my new suggestion please: https://github.com/bigcode-project/Megatron-LM/pull/33/files#r1140658122

Reply:
Awesome, thank you @NouamaneTazi !